### DISCUSSION TITLE
Shopware-CLI as official tooling for Shopware 6

### DISCUSSION THREAD

**Discussion by shyim** [shopware/platform#5307] [RFC / Proposal]
## Current State

Shopware offers some [scripts using the Flex recipe](https://github.com/shopware/recipes/tree/main/shopware/platform/6.6/bin) to the user to build/watch the storefront/administration. These scripts are written in Bash and require additional CLI tools like: `jq` installed by the user and are hard to write compatible for multiple Shopware versions.
Additionally, in `shopware/shopware`, we have additional scripts inside `composer.json` to replicate the same commands, and these both sources are drifting away or make confusion

## Desired state

Update current scripts to use the watch and build command of shopware-cli if installed, otherwise fallback to the current implementation.

For the next major, remove all `watch-*.sh` and `build-*.sh` scripts from `bin` folder so they will explicitly call `shopware-cli` and change the documentation accordingly.

## Changes needed in shopware-cli

Right now shopware-cli discovers automatically all plugins / apps that need to be built. For that it does not consider if they are active in the installation. This needs to be changed to match the current flow of the bash scripts.

https://github.com/FriendsOfShopware/shopware-cli/issues/427

## Outlook

The project shopware-cli will move soon from FriendsOfShopware to an official Shopware project, to integrate it better in the Shopware Ecosystem like OAuth login for Shopware Account, improving code review and more things

**Comment by wlwwt** [comment]
I can honestly understand the thinking behind it. We don't like the mixing of CLI and BASH scripts either.
Nevertheless, I don't think it's a good development. 

In my opinion, Core/Admin/Storefront are currently far too closely linked. Bringing all the storefront commands into the CLI now brings further dependencies with it.

I would actually like to see the build process of the storefront in particular made a little more open. In the end, it's basically just HTML (TWIG), CSS and Javascript files (and that is not meant to be derogatory). These can all remain in the storefront.
You are welcome to include the build process of the storefront in the production template as a best practice. All dependencies (such as plugins, etc.) can remain in a .json file. 
Why not even build the storefront without Shopware 6 installed? In our projects, it is always the templating effort that is the highest. Please provide us with a very simple toolset here. That would be a blast!

**Reply by shyim** [reply]
> In my opinion, Core/Admin/Storefront are currently far too closely linked. Bringing all the storefront commands into the CLI now brings further dependencies with it.

Can you explain further what the problem here is? The admin and storefront have different commands.

> I would actually like to see the build process of the storefront in particular made a little more open. In the end, it's basically just HTML (TWIG), CSS, and Javascript files (and that is not meant to be derogatory). These can all remain in the storefront.
You are welcome to include the build process of the storefront in the production template as a best practice. All dependencies (such as plugins, etc.) can remain in a .json file.
Why not even build the storefront without Shopware 6 installed? In our projects, it is always the templating effort that is the highest. Please provide us with a very simple toolset here. That would be a blast!

This was particularly addressed in the Future of Storefront to eliminate SCSS / Webpack Bundling. You can already build your custom tooling around it; you just need to put a JS file into the plugin's dist, and Shopware will pick it up. Shopware-CLI also has the option to bundle it using ESBuild; therefore, no webpack, node/npm, or Shopware itself is required.

**Comment by 0x4** [comment]
sounds good. 

what I also would like to add in this context is a flag --no-webpack (or whatever) that don't runs the node builds everytime. we always want to avoid the whole node builds in the deploy pipelines if not needed. so we currently have to call some single calls of the build-xx to achieve the same logic - because the js/css of the plugins in our case is already in git.

but maybe this is also the wrong place... and the right thing would be add a more clear documentation which scripts needs to run on every deploy

**Reply by shyim** [reply]
Shopware CLI has already something integrated which works in that way, we use that for `shopware-cli project ci` command. 
We can easily add a flag for this. I am going to add this on my bucket list

**Comment by keulinho** [comment]
Are there any more concrete benefits of switching to the CLI, such as faster build times or the ability to use a different toolchain?

How is this connected to the strategic moves in the storefront and admin components? E.g., storefront wanting to [get rid of theme compilation at runtime](https://github.com/shopware/shopware/discussions/4856#:~:text=in%20the%20future.-,Getting%20rid%20of%20the%20SCSS%20theme%20compilation,-First%2C%20we%20would), with that, the build scripts for storefront and plugins could differ even more depending on whether they want to use SCSS or not, would the plan be to have a unified opinionated build setup in the CLI in that case? 
What impact would a move from Webpack to Vite as the bundler for the administration have here? 

With that in mind, I guess in the future, some changes will be needed to how the build process behaves under the hood. I don't see that reflected in the discussion here because, as a first naive opinion, I would expect doing those changes in the external CLI that needs to be compatible with multiple/all shopware versions will be more complicated in the long run than having the build scripts as part of the project and versioned alongside the code.

Are there alternatives being considered, e.g. unifying the scripts in the repositories?

**Reply by shyim** [reply]
At the end, the CLI kinda executes `npm install` and `npm run dev.` If the npm scripts are kept the same, they can do whatever they want.

If it completely changes, we can make an if condition inside the cli to do other things. But in the end, the end-user uses the same command and doesn't have to 

>  I would expect doing those changes in the external CLI that must be compatible with multiple/all shopware versions will be more complicated in the long run than having the build scripts as part of the project and versioned alongside the code.

Yes, the building should not change so often, and it cannot also because anyone relies on webpack configuration in their extensions. The future here should be maybe to move people more into ESBuild / building outside of Shopware without dependency on Shopware itself and consider building admin/storefront to an only platform repository topic.

**Comment by JoshuaBehrens** [comment]
I am looking at it by means of a project maintainer not a plugin developer. Here are some thoughts and situations I was running into with shopware-cli. Maybe this is helpful input to determine and evaluate your approach.

---

I find many commands useful and beneficial like the more dynamic JS watch, that can also proxy an external system or the merging of snippet files. I am also a huge fan of the database anonymization tooling. Although I do not like, that it requires a running database server but the approach to have a configuration file, that each project can have to behave differently is neat.

---

So far the bash scripts were doing their job well enough for us. There was no real need to change from bash. Everyone had to know (ba)sh anyways to work on the CLI so it was not a language, that is too far off the daily biz:
* It is versioned in the project
* can be changed for each project
* runs basically everywhere because bash is widespread and is a well-ripened scripting language

With this in mind, there does not seem to be a need of a god-tool to do "everything". The symfony-cli and shopware-cli both have different downsides, that are really the opposite of bash:
* Cannot be versioned itself as it is a different binary for each OS
* What can be versioned would be a download script (which we do in our makefiles; similar to phive approach)
* It is not easy to contribute as it is written in Go, which might be widespread but is not part of the daily biz for any PHP dev. So the contribution rate is lower and therefore depends on developers, that push Go further although we are in a PHP+JS stack.
* It is not easy to adapt per project as Go is compiled
* You cannot rely on a "globally" installed shopware-cli as it will mismatch with requirements for the project. I do not care if hosters provide the binary. I depend on it being installed everywhere and it might not always be the same version (dev setup, CI, hosting). This does not account for bash as it is not containing the logic but the tools to build the logic

---

shopware-cli for example is not easy to understand and debug in a scenario like this e.g.:
1. I go `./dev-ops/bin/shopware-cli project admin-build .` and it builds assets
2. I go `./dev-ops/bin/shopware-cli project admin-watch .` and it does not build assets with this error:
```
Symfony\Component\Config\Exception\FileLocatorFileNotFoundException^ {#5886
  #message: "The file "./vendor/heptacom/project-theme//src/Resources/config/services.xml" does not exist (in: "./vendor/heptacom/project-theme//src")."
  #code: 0
  #file: "./vendor/symfony/config/FileLocator.php"
  #line: 73
  -paths: array:1 [
    0 => "./vendor/heptacom/project-theme//src/./vendor/heptacom/project-theme//src/Resources/config/services.xml"
  ]
  trace: {
    ./vendor/symfony/config/FileLocator.php:73 { …}
    ./vendor/symfony/dependency-injection/Loader/XmlFileLoader.php:46 { …}
    ./vendor/symfony/config/Loader/DelegatingLoader.php:37 { …}
    ./vendor/shopware/core/Framework/Bundle.php:160 { …}
    ./vendor/shopware/core/Framework/Bundle.php:33 { …}
    ./vendor/symfony/http-kernel/Kernel.php:629 { …}
    ./vendor/symfony/http-kernel/Kernel.php:606 { …}
    ./vendor/symfony/http-kernel/Kernel.php:494 { …}
    ./vendor/shopware/core/Kernel.php:152 { …}
    ./bin/console:57 {
      {closure}^
      › $application = new Application($kernel);
      › $kernel->boot();
      › 
    }
    ./vendor/symfony/runtime/Resolver/DebugClosureResolver.php:25 { …}
    ./vendor/autoload_runtime.php:24 { …}
    ./bin/console:17 { …}
  }
}
2024-11-02T14:42:11+00:00 [critical] Uncaught Exception: The file "./vendor/heptacom/project-theme//src/Resources/config/services.xml" does not exist (in: "./vendor/heptacom/project-theme//src")
```
I can clearly tell, that some path concatenation with relative paths are broken. And maybe I am confused, why it is broken on my end because it never was broken that way. I cannot just find out what it tries to do. So I just do not use it in that situation. Sad.

---

We tried your docker base and run into an issue, that one cannot build the container. Ok why? Apparently the plugin order was not correct on installation. So we use our scripts, that one would use anyway needs to be used. So for this we had to understand how the docker setup worked and after that, see that shopware-cli is used and end up infront of a blackbox again. So what do we do? We replace it because there was no way to configure it as plugin installation order seem to be a thing we need to do.

```Dockerfile
RUN sed -i "s/ install_all_plugins/ composer run post-update-cmd/g" /usr/local/shopware/setup_6.6.x
```

We actively ditched shopware-cli which would've done a very well job eventually just because one step failed. This is a real bummer for the tooling it provides.

---

You do not have to depend on jq. There is nothing stopping you to use the languages you expect to exist in the project anyways

```php
#!/usr/bin/env php
<?php

foreach (json_decode(file_get_contents(__DIR__ . '/../var/plugins.json')) as $plugin) {
    echo "{$plugin->basePath}{$plugin->administration->path}\t{$plugin->technicalName}" . PHP_EOL;
}
```

Then use the output to process with `sed`, `awk`, `tr`, bash substring utilities or whatever else.

---

> are hard to write compatible for multiple Shopware versions

Then just don't try to build a solution, that fits every Shopware version. It does not work for plugins to work for every major or even minor version, the API has been changed a lot in 6.2, 6.3 and 6.4, apps also have to fight with different templates in each version and I do not want to start with Twig scopes!!! that make compatibility a hell. Even the security plugin is a new version each major update since 6.4. I would not mind, that your tools expect to know which version we are in (by reading the composer.lock) and just make something, that works well for each version.

---

I would like, that you find ways to improve everyday workflows with new tools, yes. I also see the benefits, that shopware-cli already provides and it already is part of my daily biz. But a compiled god-tool in a UNIX world where you have thousands of small tools, that do just a limited-scope job (like `jq` just is a toolbox just for JSON and you need `xq` for XML, `awk` providing a string toolbox, `grep` providing a regex search, `find` a file find toolbox, ...), that are just building blocks to build your final solution seems not to be a matching way for the environments our projects eventually run in.

**Reply by JoshuaBehrens** [reply]
Just an additional "issue", that I do not find time to report "properly". I have a change in SwagCommercial and compile my storefront assets with shopware-cli and it does not recompile SwagCommercial. bin/build-storefront.sh does though

**Comment by lasomethingsomething** [comment]
The CLI moved to the /shopware GitHub org some time ago, and we plan to continue enhancing and investing in it, so I'm closing this discussion. Thanks to all who provided feedback.
